File: apps/mobile/lib/src/features/vocabulary/data/vocabulary_api_client.dart
import 'dart:convert';

import 'package:dio/dio.dart';

import '../../../core/network/api_client.dart';
import '../../../core/network/api_envelope.dart';
import '../../../core/network/api_error.dart';
import '../domain/vocabulary_card.dart';

class VocabularyApiClient {
  VocabularyApiClient({Dio? dio, ApiClient? apiClient})
    : _dio = dio ?? apiClient?.dio ?? ApiClient().dio;

  final Dio _dio;

  Future<VocabularyCardResult> createLexemeCard({
    required String lexemeId,
    String? sourceBookFingerprint,
    String? sourceBookTitle,
  }) async {
    final data = await _request(
      'POST',
      '/api/v1/vocabulary/cards',
      data: {
        'cardType': 'lexeme',
        'lexemeId': lexemeId,
        'sourceBookFingerprint': ?sourceBookFingerprint,
        'sourceBookTitle': ?sourceBookTitle,
      },
      allowExistingCardResult: true,
    );
    return VocabularyCardResult(
      card: VocabularyCard.fromJson(data),
      existing: data['_existing'] == true,
    );
  }

  Future<VocabularyCardResult> createPrivateSentenceCard({
    required String privateSurface,
    required String privateDefinition,
    String? privateContext,
    String? sourceBookFingerprint,
    String? sourceBookTitle,
  }) async {
    final data = await _request(
      'POST',
      '/api/v1/vocabulary/cards',
      data: {
        'cardType': 'private_sentence',
        'privateSurface': privateSurface,
        'privateDefinition': privateDefinition,
        'privateContext': ?privateContext,
        'sourceBookFingerprint': ?sourceBookFingerprint,
        'sourceBookTitle': ?sourceBookTitle,
      },
      allowExistingCardResult: true,
    );
    return VocabularyCardResult(
      card: VocabularyCard.fromJson(data),
      existing: data['_existing'] == true,
    );
  }

  Future<List<VocabularyCard>> listDueCards() async {
    final data = await _request('GET', '/api/v1/vocabulary/cards/due');
    final itemsJson = data['items'];
    if (itemsJson is! List) {
      throw const ApiError(
        code: 'INVALID_RESPONSE',
        message: 'Missing due vocabulary cards',
      );
    }
    return itemsJson
        .map(
          (item) =>
              VocabularyCard.fromJson(Map<String, dynamic>.from(item as Map)),
        )
        .toList(growable: false);
  }

  Future<VocabularyCard> reviewCard({
    required String cardId,
    required bool known,
    required DateTime reviewedAt,
  }) async {
    final data = await _request(
      'POST',
      '/api/v1/vocabulary/cards/$cardId/review',
      data: {
        'known': known,
        'reviewedAt': reviewedAt.toUtc().toIso8601String(),
      },
    );
    return VocabularyCard.fromJson(data);
  }

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    Map<String, Object?>? data,
    bool allowExistingCardResult = false,
  }) async {
    final response = await _dio.request<Object?>(
      path,
      data: data,
      options: Options(method: method, validateStatus: (_) => true),
    );

    final envelope = ApiEnvelope.fromJson(_decode(response.data));
    if (!envelope.success) {
      final error =
          envelope.error ??
          const ApiError(code: 'UNKNOWN_ERROR', message: 'Request failed');
      if (allowExistingCardResult && error.code == 'WORD_CARD_ALREADY_EXISTS') {
        final existing = _existingCardFrom(envelope.data);
        if (existing != null) {
          return {...existing, '_existing': true};
        }
      }
      throw _mapError(error);
    }

    final envelopeData = envelope.data;
    if (envelopeData == null) {
      throw const ApiError(
        code: 'INVALID_RESPONSE',
        message: 'Missing response data',
      );
    }
    return envelopeData;
  }

  Map<String, dynamic>? _existingCardFrom(Map<String, dynamic>? data) {
    if (data == null) {
      return null;
    }
    final existing = data['existing'];
    if (existing is Map) {
      return Map<String, dynamic>.from(existing);
    }
    final card = data['card'];
    if (card is Map) {
      return Map<String, dynamic>.from(card);
    }
    if (data['id'] is String) {
      return data;
    }
    return null;
  }

  ApiError _mapError(ApiError error) {
    return switch (error.code) {
      'WORD_CARD_ALREADY_EXISTS' => ApiError(
        code: 'VOCABULARY_CARD_ALREADY_EXISTS',
        message: error.message,
      ),
      'WORD_CARD_NOT_FOUND' => ApiError(
        code: 'VOCABULARY_CARD_NOT_FOUND',
        message: error.message,
      ),
      _ => error,
    };
  }

  Object? _decode(Object? data) {
    if (data is String) {
      return jsonDecode(data);
    }
    return data;
  }
}

File: apps/web-reader/services/vocabularyApiClient.ts
import { ofetch } from 'ofetch'

import type { ApiEnvelope } from '../types/api'
import { WebApiError } from '../types/api'
import type { WebCardType, WebReviewStatus } from '../types/localData'

type Fetcher = (url: string, options: Record<string, unknown>) => Promise<unknown>

export interface VocabularyApiClientOptions {
  baseUrl?: string
  fetcher?: Fetcher
}

export interface CreateLexemeCardRequest {
  lexemeId: string
  sourceBookFingerprint?: string
  sourceBookTitle?: string
}

export interface CreatePrivateSentenceCardRequest {
  privateSurface: string
  privateDefinition: string
  privateContext?: string
  sourceBookFingerprint?: string
  sourceBookTitle?: string
}

export interface VocabularyCardResult {
  id: string
  cardType: WebCardType
  reviewStatus: WebReviewStatus
  reviewCount: number
  nextReviewAt?: string | null
}

export interface VocabularyApiClient {
  createLexemeCard(input: CreateLexemeCardRequest): Promise<VocabularyCardResult>
  createPrivateSentenceCard(input: CreatePrivateSentenceCardRequest): Promise<VocabularyCardResult>
}

export function createVocabularyApiClient(options: VocabularyApiClientOptions = {}): VocabularyApiClient {
  const baseUrl = options.baseUrl ?? ''
  const fetcher = options.fetcher ?? ofetch
  return {
    createLexemeCard(input) {
      return request<VocabularyCardResult>(fetcher, baseUrl, '/api/v1/vocabulary/cards', {
        method: 'POST',
        body: {
          cardType: 'lexeme',
          lexemeId: input.lexemeId,
          sourceBookFingerprint: input.sourceBookFingerprint,
          sourceBookTitle: input.sourceBookTitle,
        },
      })
    },
    createPrivateSentenceCard(input) {
      return request<VocabularyCardResult>(fetcher, baseUrl, '/api/v1/vocabulary/cards', {
        method: 'POST',
        body: {
          cardType: 'private_sentence',
          privateSurface: input.privateSurface,
          privateDefinition: input.privateDefinition,
          privateContext: input.privateContext,
          sourceBookFingerprint: input.sourceBookFingerprint,
          sourceBookTitle: input.sourceBookTitle,
        },
      })
    },
  }
}

async function request<T>(
  fetcher: Fetcher,
  baseUrl: string,
  path: string,
  options: Record<string, unknown>,
): Promise<T> {
  const envelope = await fetcher(`${baseUrl}${path}`, options) as ApiEnvelope<T>
  if (!envelope.success) {
    const error = envelope.error ?? { code: 'UNKNOWN_ERROR', message: 'Request failed' }
    throw new WebApiError(error.code, error.message)
  }
  if (envelope.data == null) {
    throw new WebApiError('WEB_INVALID_RESPONSE', 'Missing response data')
  }
  return envelope.data
}
